### PR TITLE
Fix: Deletes scan jobs if their pods have been deleted and logs can't be c…

### LINKED ISSUE #873: Ignore parsing logs of scan jobs have been deleted
In the configauditreports controller we should ignore scan jobs that are cached but had been deleted from etcd so we cannot parse their logs. Currently we return an error, which is causing retry, but retrying does not make much sense in this case.

https://github.com/aquasecurity/starboard/blob/v0.13.0/pkg/operator/controller/configauditreport.go#L406

### PR DESCRIPTION
**Pull request by jw-s**
Fixes #873 and for other scan types which are not mentioned in this issue

The problem which is fixed in this PR is the following:

When a scan job creates the pod which handles scans, if this pod has been deleted before starboard parses the pod's logs, these jobs are never deleted which means when the max concurrent job setting is hit, part of this threshold is eaten up by jobs which never get deleted.

Signed-off-by: Joel Whittaker-Smith <jdws.dev@gmail.com>

### CODE REVIEW

File: pkg/kube/logs.go [hunk 1]
```diff
@@ -32,7 +35,7 @@ func (r *logsReader) GetLogsByJobAndContainerName(ctx context.Context, job *batc
 		return nil, fmt.Errorf("getting pod controlled by job: %q: %w", job.Namespace+"/"+job.Name, err)
 	}
 	if pod == nil {
```
Comment (chen-keinan) [inline pkg/kube/logs.go:37]: @jw-s thank you for the contribution:
maybe we should handle both cases :
1. Job deleted from Etcd, but the job exist in controller cache
2. pod deleted from Etcd and Job exist in Etcd and cache (as appear in your PR)

Comment (jw-s) [reply pkg/kube/logs.go:37]: Hi @chen-keinan Point 1 is covered before it even gets to this stage, earlier on in the reconcile look it checks if the object exists for example: https://github.com/aquasecurity/starboard/blob/main/pkg/operator/controller/vulnerabilityreport.go#L97

Comment (chen-keinan) [reply pkg/kube/logs.go:37]: @jw-s to `reconcile loop` you mention is for the `workload`, I'm referring to the 2nd reconcile loop for `scan jobs`
where 1st we tried to get it from cache :
`log.V(1).Info("Getting job from cache")
		err := r.Client.Get(ctx, req.NamespacedName, job)
		if err != nil {
                                if errors.IsNotFound(err) {
				log.V(1).Info("Ignoring cached job that must have been deleted")
				return ctrl.Result{}, nil
			}`
and only afterwards we get tries to get the job from Etcd :
`
func (r *logsReader) getPodByJob(ctx context.Context, job *batchv1.Job) (*corev1.Pod, error) {
	refreshedJob, err := r.clientset.BatchV1().Jobs(job.Namespace).Get(ctx, job.Name, metav1.GetOptions{})
	`

Comment (chen-keinan) [reply pkg/kube/logs.go:37]: @jw-s It will be great if you could add the check for `Job not found`,  referring to other PR opened on same issue #950

Comment (jw-s) [reply pkg/kube/logs.go:37]: @chen-keinan The new commit checks if the job isn't found and returns nil causing the controller to no requeue.

**Review by danielpacak — CHANGES_REQUESTED** [review]
Thank you for working on this enhancement @jw-s Overall the controllers code looks good to me. 

The only thing I'm not comfortable with are integration tests that you wrote for "relatively rare" corner case when the operator's cache cannot catch up with K8s API Server state and parse logs.

What's more, I'm afraid that the integration tests became flaky because we cannot reliably control Pods managed by JobController, which is part of K8s controller-manager. (See failed PR checks.)

To sum up, I'd just add unit tests whenever possible (and valuable). Let's keep integration (~ e2e) test nice and lean so we can afford running them on each PR.

**Comment by jw-s** [comment]
> Thank you for working on this enhancement @jw-s Overall the controllers code looks good to me.
> 
> The only thing I'm not comfortable with are integration tests that you wrote for "relatively rare" corner case when the operator's cache cannot catch up with K8s API Server state and parse logs.
> 
> What's more, I'm afraid that the integration tests might became flaky because we cannot reliably control Pods managed by JobController, which is part of K8s controller-manager. (See failed PR checks.)
> 
> To sum up, I'd just add unit tests whenever possible to keep integration (~ e2e) test nice and lean so we can afford running them on each PR. See this nice and detailed explanation why I think we should not validate everything at the integration tests phase [martinfowler.com/articles/practical-test-pyramid.html](https://martinfowler.com/articles/practical-test-pyramid.html).

Yeah I completely agree, this area is definitely flake provoking and i'm happy to remove these tests. Do we have an area where we can add unit tests to cover these cases? I couldn't seem to find it.

**Comment by danielpacak** [comment]
> > Thank you for working on this enhancement @jw-s Overall the controllers code looks good to me.
> > The only thing I'm not comfortable with are integration tests that you wrote for "relatively rare" corner case when the operator's cache cannot catch up with K8s API Server state and parse logs.
> > What's more, I'm afraid that the integration tests might became flaky because we cannot reliably control Pods managed by JobController, which is part of K8s controller-manager. (See failed PR checks.)
> > To sum up, I'd just add unit tests whenever possible to keep integration (~ e2e) test nice and lean so we can afford running them on each PR. See this nice and detailed explanation why I think we should not validate everything at the integration tests phase [martinfowler.com/articles/practical-test-pyramid.html](https://martinfowler.com/articles/practical-test-pyramid.html).
> 
> Yeah I completely agree, this area is definitely flake provoking and i'm happy to remove these tests (Though the failure is not because of a flake but actually something else). Do we have an area where we can add unit tests to cover these cases? I couldn't seem to find it.

We'd have to add _test.go files to implement unit tests for reconcilers. So far we only have integration tests for controllers. If this requires too much effort I'm also happy to approve this PR without very strict requirement for unit testing.

**Review by danielpacak — APPROVED** [review]
Thank you @jw-s for the contribution. The code LGTM!
